### PR TITLE
Add timer matching tests and random bits and bobs

### PR DESCRIPTION
**Pull request by adambuttrick**

### CODE REVIEW

File: matching-tests/ensemble/timer.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        self.execution_times = []
+
+    def __enter__(self):
+        self.start = time.perf_counter()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        end = time.perf_counter()
+        self.execution_times.append(end - self.start)
+
+    def get_stats(self, n=10):
```
Comment (dtkaczyk) [inline matching-tests/ensemble/timer.py]: `n` doesn't seem to be used anywhere in the function.

File: matching-tests/ensemble/timer.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import csv
```
Comment (dtkaczyk) [inline matching-tests/ensemble/timer.py]: This files seems to exist in multiple copies in different subdirectories. Is there a way to have only one copy somewhere and import it from there?

File: matching-tests/ensemble/timer.py (hunk 1 above)
Comment (adambuttrick) [reply matching-tests/ensemble/timer.py]: Corrected.

File: matching-tests/ensemble/timer.py (hunk 2 above)
Comment (adambuttrick) [reply matching-tests/ensemble/timer.py]: Corrected.

**Review by dtkaczyk — APPROVED** [review]

**Pull request merged**